File: goopylib/easing.py
from typing import Callable

EASING_TYPE = Callable[[float], float]


# noinspection PyUnresolvedReferences
def ease_linear() -> EASING_TYPE:
    r"""Returns a linear easing function

    .. math::

        \text{ease\_linear}(t) = t \qquad \footnotesize{0 \leq t \leq 1}

    .. image:: ../../_static/easing/linear.png
        :width: 33%
        :align: center

    ----------

    Raises:
        TypeError: easing argument must be a :obj:`float`
        ValueError: easing argument must be between 0 and 1 (inclusive)

    Examples:

        .. code-block:: python

            >>> ease = gp.EaseLinear()
            >>> ease(0.4)
            0.4
    """


# noinspection PyUnusedLocal, PyUnresolvedReferences
def ease_poly(power: float = 6) -> EASING_TYPE:
    r"""Returns a polynomial easing function

    .. math::

        \text{ease\_easing}(t) = t^{\text{power}} \qquad \footnotesize{0 \leq t \leq 1}

    Args:
        power (float) -> EASING_TYPE: The exponent defining the easing. Defaults to ``6``.

    Variants:

        .. function:: ease_poly_in()
        .. function:: ease_poly_out()

    .. image:: ../../_static/easing/polynomial.png

    ----------

    Raises:
        TypeError: easing argument must be a :obj:`float`
        ValueError: easing argument must be between 0 and 1 (inclusive)

    Examples:

        .. code-block:: python

            >>> ease = gp.ease_poly()
            >>> ease(0.5)
            0.015625
    """


# noinspection PyUnusedLocal
def ease_poly_in(power: float = 6) -> EASING_TYPE:
    pass


# noinspection PyUnusedLocal
def ease_poly_out(power: float = 6) -> EASING_TYPE:
    pass


# noinspection PyUnresolvedReferences
def ease_quad() -> EASING_TYPE:
    r"""Returns a quadratic easing function

    .. math::

        \text{ease\_quad}(t) = t^{2} \qquad \footnotesize{0 \leq t \leq 1}

    Variants:

        .. function:: ease_quad_in()
        .. function:: ease_quad_out()

    .. image:: ../../_static/easing/quadratic.png

    ----------

    Raises:
        TypeError: easing argument must be a :obj:`float`
        ValueError: easing argument must be between 0 and 1 (inclusive)

    Examples:

        .. code-block:: python

            >>> ease = gp.ease_quad()
            >>> ease(0.5)
            0.025
    """


def ease_quad_in() -> EASING_TYPE:
    pass


def ease_quad_out() -> EASING_TYPE:
    pass


# noinspection PyUnresolvedReferences
def ease_cubic() -> EASING_TYPE:
    r"""Returns a cubic easing function

    .. math::

        \text{ease\_cubic}(t) = t^{3} \qquad \footnotesize{0 \leq t \leq 1}

    Variants:

        .. function:: ease_cubic_in()
        .. function:: ease_cubic_out()

    .. image:: ../../_static/easing/cubic.png

    ----------

    Raises:
        TypeError: easing argument must be a :obj:`float`
        ValueError: easing argument must be between 0 and 1 (inclusive)

    Examples:

        .. code-block:: python

            >>> ease = gp.ease_cubic()
            >>> ease(0.5)
            0.0125
    """


def ease_cubic_in() -> EASING_TYPE:
    pass


def ease_cubic_out() -> EASING_TYPE:
    pass


# noinspection PyUnresolvedReferences
def ease_quart() -> EASING_TYPE:
    r"""Returns a quartic easing function

    .. math::

        \text{ease\_quart}(t) = t^{4} \qquad \footnotesize{0 \leq t \leq 1}

    Variants:

        .. function:: ease_quart_in()
        .. function:: ease_quart_out()

    .. image:: ../../_static/easing/quartic.png

    ----------

    Raises:
        TypeError: easing argument must be a :obj:`float`
        ValueError: easing argument must be between 0 and 1 (inclusive)

    Examples:

        .. code-block:: python

            >>> ease = gp.ease_quart()
            >>> ease(0.5)
            0.0625
    """


def ease_quart_in() -> EASING_TYPE:
    pass


def ease_quart_out() -> EASING_TYPE:
    pass


# noinspection PyUnresolvedReferences
def ease_quint() -> EASING_TYPE:
    r"""Returns a quintic easing function

    .. math::

        \text{ease\_quint}(t) = t^{5} \qquad \footnotesize{0 \leq t \leq 1}

    Variants:

        .. function:: ease_quint_in()
        .. function:: ease_quint_out()

    .. image:: ../../_static/easing/quintic.png

    ----------

    Raises:
        TypeError: easing argument must be a :obj:`float`
        ValueError: easing argument must be between 0 and 1 (inclusive)

    Examples:

        .. code-block:: python

            >>> ease = gp.ease_quint()
            >>> ease(0.5)
            0.03125
    """


def ease_quint_in() -> EASING_TYPE:
    pass


def ease_quint_out() -> EASING_TYPE:
    pass


# noinspection PyUnresolvedReferences
def ease_sin() -> EASING_TYPE:
    r"""Returns a sine easing function

    .. math::

        \text{ease\_sin}(t) = 1 - \text{cos}(\frac{\pi}{2}t) \qquad \footnotesize{0 \leq t \leq 1}

    Variants:

        .. function:: ease_sin_in()
        .. function:: ease_sin_out()

    .. image:: ../../_static/easing/sine.png

    ----------

    Raises:
        TypeError: easing argument must be a :obj:`float`
        ValueError: easing argument must be between 0 and 1 (inclusive)

    Examples:

        .. code-block:: python

            >>> ease = gp.ease_sin()
            >>> ease(0.5)
            0.29289381
    """


def ease_sin_in() -> EASING_TYPE:
    pass


def ease_sin_out() -> EASING_TYPE:
    pass


# noinspection PyUnusedLocal, PyUnresolvedReferences
def ease_exp(base: float = 2.718281828459045) -> EASING_TYPE:
    r"""Returns an exponential easing function

    .. math::

        \text{ease\_exp}(t) = \text{base}^{10(t - 1)} \qquad \footnotesize{0 \leq t \leq 1}

    Args:
        base (float) -> EASING_TYPE: The base of the exponential function. Defaults to ``e``.

    Variants:

        .. function:: ease_exp_in()
        .. function:: ease_exp_out()

    .. image:: ../../_static/easing/exponential.png

    ----------

    Raises:
        TypeError: easing argument must be a :obj:`float`
        ValueError: easing argument must be between 0 and 1 (inclusive)

    Examples:

        .. code-block:: python

            >>> ease = gp.ease_exp()
            >>> ease(0.2)
            0.0012393762590363622
    """


# noinspection PyUnusedLocal
def ease_exp_in(base: float = 2.718281828459045) -> EASING_TYPE:
    pass


# noinspection PyUnusedLocal
def ease_exp_out(base: float = 2.718281828459045) -> EASING_TYPE:
    pass


# noinspection PyUnresolvedReferences
def ease_circle() -> EASING_TYPE:
    r"""Returns a circular easing function

    .. math::

        \text{ease\_circle}(t) = 1 - \sqrt{1 - t^2} \qquad \footnotesize{0 \leq t \leq 1}

    Variants:

        .. function:: ease_circle_in()
        .. function:: ease_circle_out()

    .. image:: ../../_static/easing/circle.png

    ----------

    Raises:
        TypeError: easing argument must be a :obj:`float`
        ValueError: easing argument must be between 0 and 1 (inclusive)

    Examples:

        .. code-block:: python

            >>> ease = gp.ease_circle()
            >>> ease(0.2)
            0.04174243286252022
    """


def ease_circle_in() -> EASING_TYPE:
    pass


def ease_circle_out() -> EASING_TYPE:
    pass


# noinspection PyUnusedLocal, PyUnresolvedReferences
def ease_back(factor: float = 1.70158) -> EASING_TYPE:
    r"""Returns a back easing function

    .. math::

        \text{ease\_back}(t) = (\text{factor} + 1)\sdot t^3 - \text{factor}\sdot t^2 \qquad \footnotesize{0 \leq t
        \leq 1}

    Args:
        factor (float) -> EASING_TYPE: The scale of the easing's stretch. Defaults to ``1.70158``.

    Variants:

        .. function:: ease_back_in()
        .. function:: ease_back_out()

    .. image:: ../../_static/easing/back.png

    ----------

    Raises:
        TypeError: easing argument must be a :obj:`float`
        ValueError: easing argument must be between 0 and 1 (inclusive)

    Examples:

        .. code-block:: python

            >>> ease = gp.ease_back()
            >>> ease(0.2)
            -0.0496758408844471
    """


# noinspection PyUnusedLocal
def ease_back_in(factor: float = 1.70158) -> EASING_TYPE:
    pass


# noinspection PyUnusedLocal
def ease_back_out(factor: float = 1.70158) -> EASING_TYPE:
    pass


# noinspection PyUnusedLocal, PyUnresolvedReferences
def ease_elastic(factor: int = 3) -> EASING_TYPE:
    r"""Returns an elastic easing function

    .. math::

        \text{ease\_back}(t) = 2^{10(t - 1)}\text{sin}(\frac{t\pi}{2}(1 + 4\sdot \text{factor)}) \qquad
        \footnotesize{0 \leq t \leq 1}

    Args:
        factor (int) -> EASING_TYPE: The number of elastic oscillations. Defaults to ``3``.

    Variants:

        .. function:: ease_elastic_in()
        .. function:: ease_elastic_out()

    .. image:: ../../_static/easing/elastic.png

    ----------

    Raises:
        TypeError: easing argument must be a :obj:`float`
        ValueError: easing argument must be between 0 and 1 (inclusive)

    Examples:

        .. code-block:: python

            >>> ease = gp.ease_elastic()
            >>> ease(0.2)
            0.007430128753185272
    """


# noinspection PyUnusedLocal
def ease_elastic_in(factor: int = 3) -> EASING_TYPE:
    pass


# noinspection PyUnusedLocal
def ease_elastic_out(factor: int = 3) -> EASING_TYPE:
    pass


# noinspection PyUnusedLocal, PyUnresolvedReferences
def ease_bounce(bounces: int = 4, damping: float = 0.4) -> EASING_TYPE:
    r"""Returns a bounce easing function

        Args:
            bounces (int) -> EASING_TYPE: The number of bounces in the easing. Defaults to ``4``.
            damping (float) -> EASING_TYPE: The energy during each bounce. Defaults to ``0.4``.

        Variants:

            .. function:: ease_bounce_in()
            .. function:: ease_bounce_out()

        .. image:: ../../_static/easing/bounce.png

        ----------

        Raises:
            TypeError: easing argument must be a :obj:`float`
            ValueError: easing argument must be between 0 and 1 (inclusive)

        Examples:

            .. code-block:: python

                >>> ease = gp.ease_bounce()
                >>> ease(0.2)
                0.025422684848308563
        """


# noinspection PyUnusedLocal
def ease_bounce_in(bounces: int = 4, damping: float = 0.4) -> EASING_TYPE:
    pass


# noinspection PyUnusedLocal
def ease_bounce_out(bounces: int = 4, damping: float = 0.4) -> EASING_TYPE:
    pass


# TODO move to C++ API
# noinspection PyUnusedLocal,PyUnresolvedReferences
def combine_easing(easing1: EASING_TYPE, easing2: EASING_TYPE) -> EASING_TYPE:
    """Combines 2 easing functions together.

    Args:
        easing1: the 1st easing function for ``0 ≤ t < 0.5``
        easing2: the 2nd easing function ``0.5 ≤ t ≤ 1``

    Raises:
        TypeError: ``easing1`` and ``easing2`` must be easing functions

    Examples:

        .. code-block:: python

            >>> ease1 = gp.ease_elastic_in()
            >>> ease2 = gp.ease_bounce_out()
            >>> ease = combine_easing(ease1, ease2)

        .. image:: ../../_static/easing/combined.png
            :width: 50%
            :align: center
    """

    if callable(easing1):
        if callable(easing2):
            def closure(t: float) -> float:
                if t < 0.5:
                    return easing1(2 * t) / 2
                else:
                    return (1 + easing2(2 * t - 1)) / 2

            return closure

        raise TypeError(f"Easing function expected, got {type(easing2)}")
    raise TypeError(f"Easing function expected, got {type(easing1)}")

File: goopylib/src/goopylib/maths/Easing.cpp
#include "Easing.h"

#if !GP_LOG_EASING
#undef GP_LOGGING_LEVEL
#endif

#include "goopylib/debug/LogMacros.h"

#if !GP_VALUE_CHECK_EASING
#undef GP_VALUE_CHECKING
#endif

#include "goopylib/debug/Error.h"

#define EASE_IN(func, ...) func(t, ##__VA_ARGS__)
#define EASE_OUT(func, ...) (1 - func(1 - t, ##__VA_ARGS__))
#define EASE_INOUT(func, ...) (t < 0.5 ? func(2 * t, ##__VA_ARGS__) / 2 : 1 - func(2 * (1 - t), ##__VA_ARGS__) / 2)

namespace {
    void processEaseBounce(int bounces, float &damping, float &scale, std::vector<float> &zeroes,
                           std::vector<float> &derivatives) {

        GP_CHECK_GE(damping, 0, "damping value for easing must be greater than or equal to 0");
        GP_CHECK_GE(bounces, 1, "bounces value for easing must be greater than or equal to 1");

        damping = -sqrt(damping);
        zeroes.push_back(1);
        derivatives.push_back(-2);

        zeroes.push_back(-2 * damping + 1);
        derivatives.push_back(2 * (zeroes[0] - zeroes[1]) + damping * derivatives[0]);

        for (int i = 2; i < bounces; i++) {
            zeroes.push_back(damping * (zeroes[i - 2] - zeroes[i - 1]) + zeroes[i - 1]);
            derivatives.push_back(2 * (zeroes[i - 1] - zeroes[i]) + damping * derivatives[i - 1]);
        }

        scale = zeroes[bounces - 1];
    }
}

namespace gp {
    std::function<float(float)> easeLinear() {
        GP_CORE_INFO("gp::easeLinear()");

        return [](float t) {
            return t;
        };
    }

    float easePoly(float t, float power) {
        GP_CORE_TRACE("gp::easePoly(t={0}, power={1})", t, power);

        return pow(t, power);
    }

    std::function<float(float)> easePolyIn(float power) {
        GP_CORE_INFO("gp::easePolyIn(power={0})", power);

        return [power](float t) {
            return EASE_IN(easePoly, power);
        };
    }

    std::function<float(float)> easePolyOut(float power) {
        GP_CORE_INFO("gp::easePolyOut(power={0})", power);

        return [power](float t) {
            return EASE_OUT(easePoly, power);
        };
    }

    std::function<float(float)> easePolyInOut(float power) {
        GP_CORE_INFO("gp::easePolyInOut(power={0})", power);

        return [power](float t) {
            return EASE_INOUT(easePoly, power);
        };
    }

    float easeQuad(float t) {
        GP_CORE_TRACE("gp::easeQuad(t={0})", t);

        return t * t;
    }

    std::function<float(float)> easeQuadIn() {
        GP_CORE_INFO("gp::easeQuadIn()");

        return [](float t) {
            return EASE_IN(easeQuad);
        };
    }

    std::function<float(float)> easeQuadOut() {
        GP_CORE_INFO("gp::easeQuadOut()");

        return [](float t) {
            return EASE_OUT(easeQuad);
        };
    }

    std::function<float(float)> easeQuadInOut() {
        GP_CORE_INFO("gp::easeQuadInOut()");

        return [](float t) {
            return EASE_INOUT(easeQuad);
        };
    }

    float easeCubic(float t) {
        GP_CORE_TRACE("gp::easeCubic(t={0})", t);

        return t * t * t;
    }

    std::function<float(float)> easeCubicIn() {
        GP_CORE_INFO("gp::easeCubicIn()");

        return [](float t) {
            return EASE_IN(easeCubic);
        };
    }

    std::function<float(float)> easeCubicOut() {
        GP_CORE_INFO("gp::easeCubicOut()");

        return [](float t) {
            return EASE_OUT(easeCubic);
        };
    }

    std::function<float(float)> easeCubicInOut() {
        GP_CORE_INFO("gp::easeCubicInOut()");

        return [](float t) {
            return EASE_INOUT(easeCubic);
        };
    }

    float easeQuart(float t) {
        GP_CORE_TRACE("gp::easeQuart(t={0})", t);

        return t * t * t * t;
    }

    std::function<float(float)> easeQuartIn() {
        GP_CORE_INFO("gp::easeQuartIn()");

        return [](float t) {
            return EASE_IN(easeQuart);
        };
    }

    std::function<float(float)> easeQuartOut() {
        GP_CORE_INFO("gp::easeQuartOut()");

        return [](float t) {
            return EASE_OUT(easeQuart);
        };
    }

    std::function<float(float)> easeQuartInOut() {
        GP_CORE_INFO("gp::easeQuartInOut()");

        return [](float t) {
            return EASE_INOUT(easeQuart);
        };
    }

    float easeQuint(float t) {
        GP_CORE_TRACE("gp::easeQuint(t={0})", t);

        return t * t * t * t * t;
    }

    std::function<float(float)> easeQuintIn() {
        GP_CORE_INFO("gp::easeQuintIn()");

        return [](float t) {
            return EASE_IN(easeQuint);
        };
    }

    std::function<float(float)> easeQuintOut() {
        GP_CORE_INFO("gp::easeQuintOut()");

        return [](float t) {
            return EASE_OUT(easeQuint);
        };
    }

    std::function<float(float)> easeQuintInOut() {
        GP_CORE_INFO("gp::easeQuintInOut()");

        return [](float t) {
            return EASE_INOUT(easeQuint);
        };
    }

    float easeCircle(float t) {
        GP_CORE_TRACE("gp::easeCircle(t={0})", t);

        return 1 - sqrt(1 - (float) pow(t, 2));
    }

    std::function<float(float)> easeCircleIn() {
        GP_CORE_INFO("gp::easeCircleIn()");

        return [](float t) {
            return EASE_IN(easeCircle);
        };
    }

    std::function<float(float)> easeCircleOut() {
        GP_CORE_INFO("gp::easeCircleOut()");

        return [](float t) {
            return EASE_OUT(easeCircle);
        };
    }

    std::function<float(float)> easeCircleInOut() {
        GP_CORE_INFO("gp::easeCircleInOut()");

        return [](float t) {
            return EASE_INOUT(easeCircle);
        };
    }

    float easeSin(float t) {
        GP_CORE_TRACE("gp::easeSin(t={0})", t);

        return 1 - (float) cos(t * 1.57079632679);
    }

    std::function<float(float)> easeSinIn() {
        GP_CORE_INFO("gp::easeSinIn()");

        return [](float t) {
            return EASE_IN(easeSin);
        };
    }

    std::function<float(float)> easeSinOut() {
        GP_CORE_INFO("gp::easeSinOut()");

        return [](float t) {
            return EASE_OUT(easeSin);
        };
    }

    std::function<float(float)> easeSinInOut() {
        GP_CORE_INFO("gp::easeSinInOut()");

        return [](float t) {
            return EASE_INOUT(easeSin);
        };
    }

    float easeExp(float t, float base) {
        GP_CORE_TRACE("gp::easeExp(t={0}, base={1})", t, base);

        return pow(base, 10 * t - 10);
    }

    std::function<float(float)> easeExpIn(float base) {
        GP_CORE_INFO("gp::easeExpIn(base={0})", base);

        return [base](float t) {
            return EASE_IN(easeExp, base);
        };
    }

    std::function<float(float)> easeExpOut(float base) {
        GP_CORE_INFO("gp::easeExpOut(base={0})", base);

        return [base](float t) {
            return EASE_OUT(easeExp, base);
        };
    }

    std::function<float(float)> easeExpInOut(float base) {
        GP_CORE_INFO("gp::easeExpInOut(base={0})", base);

        return [base](float t) {
            return EASE_INOUT(easeExp, base);
        };
    }

    float easeBack(float t, float factor) {
        GP_CORE_TRACE("gp::easeBack(t={0}, factor={1})", t, factor);

        return ((factor + 1) * t * t * t - factor * t * t);
    }

    std::function<float(float)> easeBackIn(float factor) {
        GP_CORE_INFO("gp::easeBackIn(factor={0})", factor);

        return [factor](float t) {
            return EASE_IN(easeBack, factor);
        };
    }

    std::function<float(float)> easeBackOut(float factor) {
        GP_CORE_INFO("gp::easeBackOut(factor={0})", factor);

        return [factor](float t) {
            return EASE_OUT(easeBack, factor);
        };
    }

    std::function<float(float)> easeBackInOut(float factor) {
        GP_CORE_INFO("gp::easeBackInOut(factor={0})", factor);

        return [factor](float t) {
            return EASE_INOUT(easeBack, factor);
        };
    }

    float easeElastic(float t, int factor) {
        GP_CORE_TRACE("gp::easeElastic(t={0}, factor={1})", t, factor);

        return (float) pow(2, 10 * (t - 1)) * (float) std::sin(1.5707963267 * t * (1 + 4 * factor));
    }

    std::function<float(float)> easeElasticIn(int factor) {
        GP_CORE_INFO("gp::easeElasticIn(factor={0})", factor);

        return [factor](float t) {
            return EASE_IN(easeElastic, factor);
        };
    }

    std::function<float(float)> easeElasticOut(int factor) {
        GP_CORE_INFO("gp::easeElasticOut(factor={0})", factor);

        return [factor](float t) {
            return EASE_OUT(easeElastic, factor);
        };
    }

    std::function<float(float)> easeElasticInOut(int factor) {
        GP_CORE_INFO("gp::easeElasticInOut(factor={0})", factor);

        return [factor](float t) {
            return EASE_INOUT(easeElastic, factor);
        };
    }

    float easeBounce(float t, float damping, int bounces, float scale, const std::vector<float> &zeroes,
                     const std::vector<float> &derivatives) {
        GP_CORE_TRACE("gp::easeBounce(t={0}, bounces={1}, damping={2}, scale={3})", t, bounces, damping, scale);

        t = scale * (1 - t);
        for (int i = 0; i < bounces; i++) {
            if (zeroes[i] >= t) {
                if (i == 0) {
                    return -t * t + 1;
                }
                else {
                    t -= zeroes[i - 1];
                    return -t * t + damping * t * derivatives[i - 1];
                }
            }
        }
        return 1;
    }

    std::function<float(float)> easeBounceIn(int bounces, float damping) {
        GP_CORE_INFO("gp::easeBounceIn(bounces={0}, damping={1})", bounces, damping);

        std::vector<float> zeroes, derivatives;
        float scale;
        processEaseBounce(bounces, damping, scale, zeroes, derivatives);

        return [damping, bounces, scale, zeroes, derivatives](float t) {
            return EASE_IN(easeBounce, damping, bounces, scale, zeroes, derivatives);
        };
    }

    std::function<float(float)> easeBounceOut(int bounces, float damping) {
        GP_CORE_INFO("gp::easeBounceOut(bounces={0}, damping={1})", bounces, damping);

        std::vector<float> zeroes, derivatives;
        float scale;
        processEaseBounce(bounces, damping, scale, zeroes, derivatives);

        return [damping, bounces, scale, zeroes, derivatives](float t) {
            return EASE_OUT(easeBounce, damping, bounces, scale, zeroes, derivatives);
        };
    }

    std::function<float(float)> easeBounceInOut(int bounces, float damping) {
        GP_CORE_INFO("gp::easeBounceInOut(bounces={0}, damping={1})", bounces, damping);

        std::vector<float> zeroes, derivatives;
        float scale;
        processEaseBounce(bounces, damping, scale, zeroes, derivatives);

        return [damping, bounces, scale, zeroes, derivatives](float t) {
            return EASE_INOUT(easeBounce, damping, bounces, scale, zeroes, derivatives);
        };
    }

    std::function<float(float)> combineEasing(const std::function<float(float)> &func1,
                                              const std::function<float(float)> &func2) {
        GP_CORE_INFO("gp::combineEasing()");

        return [func1, func2](float t) {
            if (t < 0.5) {
                return func1(2 * t) / 2;
            }
            return (1 - func2(2 * t - 1)) / 2;
        };
    }
}
